Refactor cox regression functions to fix table structure

edelarua opened the pull request:
Closes #841 

Downstream updates needed:
- `teal.modules.clinical` https://github.com/insightsengineering/teal.modules.clinical/pull/756
- `tlg-catalog` https://github.com/insightsengineering/tlg-catalog/pull/36
- `scda.test` https://github.com/insightsengineering/scda.test/pull/40

edelarua commented on tests/testthat/_snaps/summarize_coxreg.md:
```diff
@@ -0,0 +1,232 @@
+    Code
+      res
+    Output
+                                            Hazard Ratio      95% CI      p-value
+      ———————————————————————————————————————————————————————————————————————————
+      Treatment:                                                                 
+        ARM (reference = 1)                                                      
+          2                                     0.60       (0.40, 0.88)   0.0100 
+      Covariate:                                                                 
+        Age                                                                      
+          All                                   1.01       (0.99, 1.03)   0.2097 
+        A Covariate Label (reference = 1)                                 <0.0001
+          2                                     0.47       (0.29, 0.75)   0.0017 
+          3                                     0.31       (0.18, 0.51)   <0.0001
+          4                                     0.18       (0.10, 0.33)   <0.0001
+        Sex (F/M) (reference = F)                                                
+          M                                     1.31       (0.89, 1.92)   0.1665 
```
Note differences in multivar output:
- "Treatment" and "Covariate" labels
- numeric covariate printing (2 lines, "All" label)

edelarua commented on tests/testthat/_snaps/summarize_coxreg.md:
```diff
@@ -0,0 +1,232 @@
+    Code
+      res
+    Output
+                             n    Hazard Ratio      95% CI      p-value   Interaction p-value
+      ———————————————————————————————————————————————————————————————————————————————————————
+      Treatment:                                                                             
+        2 vs control (1)    340       0.64       (0.43, 0.94)   0.0242                       
+      Covariate:                                                                             
+        Age                 340                                                 0.1757       
+          40                          0.62       (0.42, 0.93)                                
+        A Covariate Label   340                                                 0.9883       
+          1                           0.63       (0.35, 1.14)                                
+          2                           0.58       (0.27, 1.26)                                
+          3                           0.55       (0.22, 1.35)                                
+          4                           0.69       (0.23, 2.07)                                
+        Sex (F/M)           340                                                 0.7759       
+          F                           0.67       (0.36, 1.22)                                
+          M                           0.60       (0.36, 0.99)                                
```
Note differences in univar output:
- Indentation

Melkiades commented on R/argument_convention.R:
```diff
@@ -24,6 +24,7 @@
 #'   by a statistics function.
 #' @param .spl_context (`data.frame`)\cr gives information about ancestor split states
 #'   that is passed by `rtables`.
+#' @param .na_str (`character`)\cr string that should be displayed when a value is missing. Defaults to `NA`.
```
is this similar to `na_level`?

edelarua replied:
Yes - I'll switch it to that for consistency with other functions!

Melkiades commented on R/summarize_coxreg.R:
```diff
@@ -0,0 +1,370 @@
+#'   * `ci`: confidence interval
+#'   * `pval`: p-value of the treatment effect
+#'   * `pval_inter`: p-value of the interaction effect between the treatment and the covariate (univariable only)
+#' @param .which_vars (`character`)\cr which rows should statistics be returned for from the given model.
```
bah this is up to you. It is fine

Melkiades commented on R/summarize_coxreg.R:
```diff
@@ -0,0 +1,370 @@
+#' u2_variables <- list(time = "TIME", event = "STATUS", covariates = c("COVAR1", "COVAR2"))
+#' univar_covs_model <- fit_coxreg_univar(variables = u2_variables, data = dta_bladder)
+#' df1_covs <- broom::tidy(univar_covs_model)
+#' s_coxreg(df = df1_covs, .stats = "hr", .var_nms = c("COVAR2", "Sex (F/M)"))
```
This gives me the following:
```

$hr
$hr$`Sex (F/M) (reference = F)`
numeric(0)


$hr
$hr$M
[1] 1.3271
```

Melkiades replied:
I think this comes from the fit:

```
 df1_covs
                      term         pval                        term_label        hr       lcl       ucl level   n
COVAR1.1            COVAR1 6.935431e-09 A Covariate Label (reference = 1)                  NA        NA  <NA>    
COVAR1.2 A Covariate Label 0.0007475224                                 2  0.446633 0.2795487 0.7135825     2 340
```

is it expected? It does not seem not to have any data in input...

Melkiades replied:
ahhh ok of course, it is the reference!
```
                                      Hazard Ratio      95% CI      p-value
———————————————————————————————————————————————————————————————————————————
Covariate:                                                                 
  A Covariate Label (reference = 1)                                 <0.0001
    2                                     0.45       (0.28, 0.71)   0.0007 
    3                                     0.31       (0.19, 0.52)   <0.0001
    4                                     0.18       (0.10, 0.33)   <0.0001
  Sex (F/M) (reference = F)                                                
    M                                     1.33       (0.91, 1.94)   0.1414 
```

Melkiades replied:
But why the p-value for the reference covariate is printed out?

edelarua replied:
I'm not really sure why it's printed, but it was included in the tlg-c table (see [COXT02](https://insightsengineering.github.io/tlg-catalog/tables/efficacy/coxt02.html)) and I just left all output values as is.

Melkiades commented on R/summarize_coxreg.R:
```diff
@@ -0,0 +1,370 @@
+#'   * `pval`: p-value of the treatment effect
+#'   * `pval_inter`: p-value of the interaction effect between the treatment and the covariate (univariable only)
+#' @param .which_vars (`character`)\cr which rows should statistics be returned for from the given model.
+#'   Defaults to "all". Other options include "var_main" for main effects, "inter" for interaction effects,
+#'   and "multi_lvl" for multivariable model covariate level rows. When `.which_vars` is "all" specific
```
there is only one multi_var for summarize_coxreg atm. Maybe I would add a couple more examples with the other options

Melkiades replied:
Or maybe a bit more description here or in the examples. Like the fact that you need interaction etc..

edelarua commented:
Hi @Melkiades,

One minor change - I added model caching in which decreases time to run `summarize_coxreg` tests from ~17 seconds to ~7 seconds. I re-ran all tests and didn't find any changes, but if you want to re-test as well you can (sorry for making more changes after you already reviewed!).

I also added in a few more examples for the different `s_coxreg` options as you requested.

Melkiades commented on R/summarize_coxreg.R:
```diff
@@ -309,12 +317,16 @@ summarize_coxreg <- function(lyt,
-      extra_args = list(.stats = .stats, .formats = .formats, na_level = rep(na_level, length(.stats)))
+      extra_args = list(
+        .stats = .stats, .formats = .formats, na_level = rep(na_level, length(.stats)),
+        cache_env = replicate(length(.stats), list(env))
```
does this mean that for summarize_vars there is always caching?

Melkiades commented on R/summarize_coxreg.R:
```diff
@@ -179,12 +182,17 @@ a_coxreg <- function(df,
-    model <- fit_coxreg_multivar(variables = variables, data = df, control = control) %>% broom::tidy()
+    model <- cache_env[[cov]]
   }
+  if (!multivar && !var_main) model[, "pval_inter"] <- NA_real_
```
if it cached as such it can still be in the parenthesis as before, right?

edelarua replied:
No, since first model created (for treatment effects) will be cached and we don't want to remove _all_ `pval_inter` values, we must remove them after caching.

Melkiades commented:
> One minor change - I added model caching in which decreases time to run `summarize_coxreg` tests from ~17 seconds to ~7 seconds. I re-ran all tests and didn't find any changes, but if you want to re-test as well you can (sorry for making more changes after you already reviewed!).

No problem at all! This is a very good idea and if it is not there, it should be extended to other similar layouting a_functions. This may solve other computation issues in other functions (fit_* family e.g.)
